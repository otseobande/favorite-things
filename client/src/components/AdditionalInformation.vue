<template>
  <div class="mt-4">
    <label class="font-bold">Additional information</label>
    <span class="text-sm italic text-gray-500 ml-1">(optional)</span>
    <div class="mt-1">
      <div class="flex">
        <div class="mr-1">
          <label class="text-sm mr-16">Type</label>
        </div>
        <div class="mr-1">
          <label class="text-sm mr-40">Key</label>
        </div>
        <div class="mr-1">
          <label class="text-sm ml-3">Value</label>
        </div>
      </div>
      <div class="flex mt-2" v-for="(data, index) in metadata" :key="index">
        <div class="mr-1">
          <t-select
            name="my-input"
            v-model="data.type"
            class="type-select text-sm"
            @change="handleInputTypeChange(index)"
            :options="[
              { value: 'text', text: 'Text' },
              { value: 'number', text: 'Number' },
              { value: 'date', text: 'Date' },
              { value: 'list', text: 'List' }
            ]"
          />
        </div>
        <div class="mr-1">
          <t-input class="h-10" v-model="data.label" @change="emitUpdate"/>
        </div>
        <div class="mr-1">
          <t-input
            class="h-10"
            v-model="data.value"
            @change="emitUpdate"
            :type="data.type"
            v-if="data.type !== 'list'"
          />
          <TagsInput
            v-else
            v-model="data.value"
            @tags-updated="emitUpdate"
            add-tags-on-comma
            add-tags-on-blur
            placeholder="Add item"
          />
        </div>
        <button class="text-red-500 px-1 h-8 w-8 mt-1" @click.prevent="removeField(index)">
          <CloseIcon class="fill-current w-4"/>
        </button>
      </div>
    </div>
    <t-button size="sm" class="text-sm mt-3" @click.prevent="addField">
      Add field
    </t-button>
  </div>
</template>

<script>
import CloseIcon from './CloseIcon.vue';
import TagsInput from '@voerro/vue-tagsinput';

export default {
  props: [
    'currentMetadata',
    'editting'
  ],
  components: {
    CloseIcon,
    TagsInput
  },
  data () {
    return {
      metadata: this.currentMetadata || [
        {
          type: 'text',
          label: '',
          value: ''
        }
      ],
    }
  },
  methods: {
    emitUpdate() {
      this.$emit('dataChange', this.cleanedMetadata);
    },
    handleInputTypeChange(index) {
      if (this.metadata[index].type === 'list') {
        this.metadata[index].value = []
      } else {
        this.metadata[index].value = ''
      }
      this.emitUpdate();
    },
    addField() {
      this.metadata.push({
        type: 'text',
        label: '',
        value: ''
      })
      this.emitUpdate();
    },
    removeField(index) {
      this.metadata.splice(index, 1);
      this.emitUpdate();
    },
  },
  computed: {
    cleanedMetadata () {
      return this.metadata
        .filter(data => {
          return Boolean(data.label) && Boolean(data.value);
        })
        .map(data => {
          if (data.type === 'list' && Array.isArray(data.value)) {
            const updatedData = { ...data, value: [ ...data.value ] };

            updatedData.value = updatedData.value.map(valueObject => {
              return valueObject.value;
            });

            return updatedData;
          }

          return data;
        });
    },
  },
  mounted() {
    if (this.currentMetadata) {
      this.metadata = this.currentMetadata;
    }
  }
}
</script>

<style>

.tags-input {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  max-width: 300px;
}

.tags-input input {
  flex: 1;
  background: transparent;
  border: none;
}

.tags-input input:focus {
  outline: none;
}

.tags-input input[type="text"] {
  color: #495057;
}

.tags-input-wrapper-default {
  padding: .5rem .25rem;

  background: #fff;

  border: 1px solid transparent;
  border-radius: .25rem;
  border-color: #dbdbdb;
}

/* The tag badges & the remove icon */
.tags-input span {
  margin-right: 0.3rem;
}

.tags-input-remove {
  cursor: pointer;
  position: relative;
  display: inline-block;
  width: 0.5rem;
  height: 0.5rem;
  overflow: hidden;
}

.tags-input-remove:before, .tags-input-remove:after {
  content: '';
  position: absolute;
  width: 100%;
  top: 50%;
  left: 0;
  background: #5dc282;

  height: 2px;
  margin-top: -1px;
}

.tags-input-remove:before {
    transform: rotate(45deg);
}
.tags-input-remove:after {
    transform: rotate(-45deg);
}

/* Tag badge styles */
.tags-input-badge {
    display: inline-block;
    padding: 0.25em 0.4em;
    font-size: 75%;
    font-weight: 700;
    line-height: 1;
    text-align: center;
    white-space: nowrap;
    vertical-align: baseline;
    border-radius: 0.25rem;
}

.tags-input-badge-pill {
    padding-right: 0.6em;
    padding-left: 0.6em;
    border-radius: 10rem;
}

.tags-input-badge-selected-default {
    color: #212529;
    background-color: #f0f1f2;
}

/* Typeahead */
.typeahead-hide-btn {
    color: #999 !important;
    font-style: italic;
}

/* Typeahead - badges */
.typeahead-badges > span {
    cursor: pointer;
    margin-right: 0.3rem;
}

/* Typeahead - dropdown */
.typeahead-dropdown {
    list-style-type: none;
    padding: 0;
    margin: 0;
    position: absolute;
    width: 100%;
    z-index: 1000;
}

.typeahead-dropdown li {
    padding: .25rem 1rem;
    cursor: pointer;
}

/* Typeahead elements style/theme */
.tags-input-typeahead-item-default {
    color: #fff;
    background-color: #343a40;
}

.tags-input-typeahead-item-highlighted-default {
    color: #fff;
    background-color: #007bff;
}
</style>
