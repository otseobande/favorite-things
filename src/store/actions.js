import * as actionTypes from './actions.type';
import * as mutationTypes from './mutations.type';

import categoryService from '../services/categoryService';
import favoriteThingService from '../services/favoriteThingService';

export default {
  async [actionTypes.FETCH_CATEGORIES]({ commit }) {
    const response = await categoryService.get();

    commit(mutationTypes.SET_CATEGORIES, response.data.results)
  },

  async [actionTypes.ADD_FAVORITE_THING]({ commit }, newFavoriteThing) {
    const response = await favoriteThingService.create(newFavoriteThing);

    commit(mutationTypes.ADD_FAVORITE_THING, response.data)
  },

  async [actionTypes.UPDATE_FAVORITE_THING]({ commit }, {
    favoriteThingId,
    updatedFavoriteThingData
  }) {
    const response = await favoriteThingService.update({
      favoriteThingId,
      updatedFavoriteThingData
    })

    commit(mutationTypes.UPDATE_FAVORITE_THING, response.data);
  },

  async [actionTypes.DELETE_FAVORITE_THING]({ commit }, { favoriteThingId }) {
    await favoriteThingService.delete({
      favoriteThingId
    });

    commit(mutationTypes.DELETE_FAVORITE_THING, favoriteThingId)
  },

  async [actionTypes.ADD_DUMMY_CATEGORY]({ commit }) {
    commit(mutationTypes.ADD_CATEGORY, {
      id: 'template',
      name: '',
      favorite_things: []
    })
  },

  async [actionTypes.DELETE_DUMMY_CATEGORY]({ commit }) {
    commit(mutationTypes.DELETE_CATEGORY, 'template')
  },

  async [actionTypes.ADD_CATEGORY]({ commit }, newCategory) {
    const response = await categoryService.create(newCategory)

    commit(mutationTypes.ADD_CATEGORY, response.data)
  },

  async [actionTypes.DELETE_CATEGORY]({ commit }, { categoryId }) {
    await categoryService.delete(categoryId);

    commit(mutationTypes.DELETE_CATEGORY, categoryId);
  }
}
