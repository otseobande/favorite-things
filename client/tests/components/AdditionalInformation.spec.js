import { shallowMount, createLocalVue } from '@vue/test-utils';
import AdditionalInformation from '../../src/components/AdditionalInformation.vue';
import VueTailwind from 'vue-tailwind';
import Vuex from 'vuex';
import actions from '../../src/store/actions';
import Vuelidate from 'vuelidate';
import * as actionTypes from '../../src/store/actions.type';

jest.mock('../../src/store/actions');

const localVue = createLocalVue();

localVue.use(VueTailwind);
localVue.use(Vuex);
localVue.use(Vuelidate);

let wrapper;
const propsData = {
  category: {
    name: 'Test category',
    favorite_things: []
  },
  singularizedCategoryName: 'Test category'
}

describe('AdditionalInformation component', () => {
  beforeEach(() => {
    wrapper = shallowMount(AdditionalInformation, {
      localVue,
      store: new Vuex.Store({
        actions
      }),
      propsData,
    });
  })
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('emitUpdate method', () => {
    it('should emit dataChange event', () => {
      wrapper.vm.$emit = jest.fn()

      wrapper.vm.emitUpdate()

      expect(wrapper.vm.$emit).toBeCalledWith(
        'dataChange',
        wrapper.vm.cleanedMetadata
      );
    })
  });

  describe('handleInputTypeChange', () => {
    it('should reset metadata value for index if not list to string', () => {
      wrapper.vm.emitUpdate = jest.fn()
      wrapper.vm.handleInputTypeChange(0);

      expect(wrapper.vm.metadata[0].value).toEqual('');
      expect(wrapper.vm.emitUpdate).toBeCalled()
    });
    it('should reset metadata value for index if type is list to an array', () => {
      wrapper.vm.metadata = [
        {
          type: 'list',
          label: '',
          value: ''
        }
      ]
      wrapper.vm.emitUpdate = jest.fn()
      wrapper.vm.handleInputTypeChange(0);

      expect(wrapper.vm.metadata[0].value).toEqual([]);
      expect(wrapper.vm.emitUpdate).toBeCalled()
    });
  });

  describe('addField method', () => {
    it('should push new metadata field', () => {
      expect(wrapper.vm.metadata.length).toEqual(1)

      wrapper.vm.addField()

      expect(wrapper.vm.metadata.length).toEqual(2)
    });
    it('should emit update', () => {
      wrapper.vm.emitUpdate = jest.fn()
      wrapper.vm.addField()
      expect(wrapper.vm.emitUpdate).toBeCalled()
    })
  })

  describe('removeField method', () => {
    it('should push new metadata field', () => {
      expect(wrapper.vm.metadata.length).toEqual(1)

      wrapper.vm.removeField()

      expect(wrapper.vm.metadata.length).toEqual(0)
    });
    it('should emit update', () => {
      wrapper.vm.emitUpdate = jest.fn()
      wrapper.vm.removeField()
      expect(wrapper.vm.emitUpdate).toBeCalled()
    })
  });

  describe('cleanMetaData computed field', () => {
    it('should flatten list fields', () => {
      wrapper.vm.metadata = [
        {
          type: 'text',
          label: 'name',
          value: 'Otse'
        },
        {
          type: 'list',
          label: 'hobbies',
          value: [
            {
              key: '',
              value: 'coding'
            },
            {
              key: '',
              value: 'coding'
            }
          ]
        }
      ]

      expect(wrapper.vm.cleanedMetadata).toEqual(
        [
          {
            type: 'text',
            label: 'name',
            value: 'Otse'
          },
          {
            type: 'list',
            label: 'hobbies',
            value: ['coding', 'coding']
          }
        ]
      )
    })
  });

  it('should map currentMetadata prop to metadata state', () => {
    wrapper = shallowMount(AdditionalInformation, {
      localVue,
      store: new Vuex.Store({
        actions
      }),
      propsData: {
        ...propsData,
        currentMetadata: []
      },
    });

    expect(wrapper.vm.metadata).toEqual([])
  })
})
