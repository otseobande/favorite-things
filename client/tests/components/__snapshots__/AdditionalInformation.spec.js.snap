// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInformation component should match snapshot 1`] = `
<div class="mt-4" category="[object Object]" singularizedcategoryname="Test category"><label class="font-bold">Additional information</label> <span class="text-sm italic text-gray-500 ml-1">(optional)</span>
  <div class="mt-1">
    <div class="flex">
      <div class="mr-1"><label class="text-sm mr-16">Type</label></div>
      <div class="mr-1"><label class="text-sm mr-40">Key</label></div>
      <div class="mr-1"><label class="text-sm ml-3">Value</label></div>
    </div>
    <div class="flex mt-2">
      <div class="mr-1">
        <t-select-stub name="my-input" options="[object Object],[object Object],[object Object],[object Object]" value="text" baseclass="block appearance-none w-full border pr-8 rounded leading-tight" baseclassmultiple="block appearance-none w-full border rounded leading-tight" defaultstatusclass="bg-white" errorstatusclass="border-red-300 bg-red-100" successstatusclass="border-green-300 bg-green-100" warningstatusclass="border-yellow-400 bg-yellow-100" disabledclass="bg-gray-100 cursor-not-allowed opacity-75" arrowwrapperclass="pointer-events-none absolute inset-y-0 right-0 flex items-center px-2 text-gray-700" defaultsizeclass="p-3" largesizeclass="p-4 text-lg" smallsizeclass="p-2 text-sm" arrowclass="fill-current h-4 w-4" wrapperclass="relative" class="type-select text-sm"></t-select-stub>
      </div>
      <div class="mr-1">
        <t-input-stub value="" type="text" baseclass="border block rounded" defaultstatusclass="bg-white" errorstatusclass="border-red-300 bg-red-100" successstatusclass="border-green-300 bg-green-100" warningstatusclass="border-yellow-400 bg-yellow-100" disabledclass="bg-gray-100 cursor-not-allowed opacity-75" defaultsizeclass="p-3" largesizeclass="p-4 text-lg" smallsizeclass="p-2 text-sm" class="h-10"></t-input-stub>
      </div>
      <div class="mr-1">
        <t-input-stub value="" type="text" baseclass="border block rounded" defaultstatusclass="bg-white" errorstatusclass="border-red-300 bg-red-100" successstatusclass="border-green-300 bg-green-100" warningstatusclass="border-yellow-400 bg-yellow-100" disabledclass="bg-gray-100 cursor-not-allowed opacity-75" defaultsizeclass="p-3" largesizeclass="p-4 text-lg" smallsizeclass="p-2 text-sm" class="h-10"></t-input-stub>
      </div> <button class="text-red-500 px-1 h-8 w-8 mt-1">
        <closeicon-stub class="fill-current w-4"></closeicon-stub>
      </button>
    </div>
  </div>
  <t-button-stub tagname="button" type="button" size="sm" activeclass="router-link-active" exactactiveclass="router-link-exact-active" baseclass="border block rounded inline-flex items-center justify-center" defaultclass="bg-white border-gray-400 hover:bg-gray-100 hover:border-gray-500" primaryclass="text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600" secondaryclass="border-blue-500 text-blue-500 bg-white hover:border-blue-600 hover:text-blue-600 hover:bg-white" tertiaryclass="border block underline text-blue-500 border-transparent bg-transparent hover:text-blue-600" successclass="text-white bg-green-500 border-green-500 hover:bg-green-600 hover:border-green-600" dangerclass="text-white bg-red-500 border-red-500 hover:bg-red-600 hover:border-red-600" warningclass="text-yellow-900 bg-yellow-500 border-yellow-500 hover:bg-yellow-600 hover:border-yellow-600" disabledclass="cursor-not-allowed opacity-75" defaultsizeclass="px-6 py-3" largesizeclass="px-8 py-4 text-lg" smallsizeclass="px-4 py-2 text-sm" class="text-sm mt-3">
    Add field
  </t-button-stub>
</div>
`;
